File: mobile/lib/data/models/activity.dart
import 'package:json_annotation/json_annotation.dart';

part 'activity.g.dart';

/// Mirrors `stride_core` Activity row + Activity dataclass in
/// `frontend/src/api.ts:243`. Field names match the JSON 1:1.
@JsonSerializable()
class Activity {
  const Activity({
    required this.labelId,
    required this.sportType,
    required this.sportName,
    required this.date,
    required this.distanceM,
    required this.distanceKm,
    required this.durationS,
    required this.durationFmt,
    required this.paceFmt,
    this.name,
    this.avgPaceSKm,
    this.avgHr,
    this.maxHr,
    this.avgCadence,
    this.caloriesKcal,
    this.trainingLoad,
    this.vo2max,
    this.trainType,
    this.ascentM,
    this.aerobicEffect,
    this.anaerobicEffect,
    this.temperature,
    this.humidity,
    this.feelsLike,
    this.windSpeed,
    this.feelType,
    this.sportNote,
    this.commentary,
    this.commentaryGeneratedBy,
    this.commentaryGeneratedAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json) =>
      _$ActivityFromJson(json);

  @JsonKey(name: 'label_id')
  final String labelId;
  final String? name;
  @JsonKey(name: 'sport_type')
  final int sportType;
  @JsonKey(name: 'sport_name')
  final String sportName;
  final String date;
  @JsonKey(name: 'distance_m')
  final num distanceM;
  @JsonKey(name: 'distance_km')
  final num distanceKm;
  @JsonKey(name: 'duration_s')
  final int durationS;
  @JsonKey(name: 'duration_fmt')
  final String durationFmt;
  @JsonKey(name: 'avg_pace_s_km')
  final num? avgPaceSKm;
  @JsonKey(name: 'pace_fmt')
  final String paceFmt;
  @JsonKey(name: 'avg_hr')
  final int? avgHr;
  @JsonKey(name: 'max_hr')
  final int? maxHr;
  @JsonKey(name: 'avg_cadence')
  final int? avgCadence;
  @JsonKey(name: 'calories_kcal')
  final num? caloriesKcal;
  @JsonKey(name: 'training_load')
  final num? trainingLoad;
  final num? vo2max;
  @JsonKey(name: 'train_type')
  final String? trainType;
  @JsonKey(name: 'ascent_m')
  final num? ascentM;
  @JsonKey(name: 'aerobic_effect')
  final num? aerobicEffect;
  @JsonKey(name: 'anaerobic_effect')
  final num? anaerobicEffect;
  final num? temperature;
  final num? humidity;
  @JsonKey(name: 'feels_like')
  final num? feelsLike;
  @JsonKey(name: 'wind_speed')
  final num? windSpeed;
  @JsonKey(name: 'feel_type')
  final int? feelType;
  @JsonKey(name: 'sport_note')
  final String? sportNote;
  final String? commentary;
  @JsonKey(name: 'commentary_generated_by')
  final String? commentaryGeneratedBy;
  @JsonKey(name: 'commentary_generated_at')
  final String? commentaryGeneratedAt;

  Map<String, dynamic> toJson() => _$ActivityToJson(this);
}

@JsonSerializable()
class Lap {
  const Lap({
    required this.lapIndex,
    required this.lapType,
    required this.distanceM,
    required this.distanceKm,
    required this.durationS,
    required this.durationFmt,
    required this.paceFmt,
    this.avgPace,
    this.adjustedPace,
    this.avgHr,
    this.maxHr,
    this.avgCadence,
    this.avgPower,
    this.ascentM,
    this.descentM,
  });

  factory Lap.fromJson(Map<String, dynamic> json) => _$LapFromJson(json);

  @JsonKey(name: 'lap_index')
  final int lapIndex;
  @JsonKey(name: 'lap_type')
  final String lapType;
  @JsonKey(name: 'distance_m')
  final num distanceM;
  @JsonKey(name: 'distance_km')
  final num distanceKm;
  @JsonKey(name: 'duration_s')
  final num durationS;
  @JsonKey(name: 'duration_fmt')
  final String durationFmt;
  @JsonKey(name: 'avg_pace')
  final num? avgPace;
  @JsonKey(name: 'pace_fmt')
  final String paceFmt;
  @JsonKey(name: 'adjusted_pace')
  final num? adjustedPace;
  @JsonKey(name: 'avg_hr')
  final int? avgHr;
  @JsonKey(name: 'max_hr')
  final int? maxHr;
  @JsonKey(name: 'avg_cadence')
  final int? avgCadence;
  @JsonKey(name: 'avg_power')
  final num? avgPower;
  @JsonKey(name: 'ascent_m')
  final num? ascentM;
  @JsonKey(name: 'descent_m')
  final num? descentM;

  Map<String, dynamic> toJson() => _$LapToJson(this);
}

@JsonSerializable()
class Zone {
  const Zone({
    required this.zoneType,
    required this.zoneIndex,
    required this.rangeUnit,
    required this.durationS,
    required this.percent,
    this.rangeMin,
    this.rangeMax,
  });

  factory Zone.fromJson(Map<String, dynamic> json) => _$ZoneFromJson(json);

  @JsonKey(name: 'zone_type')
  final String zoneType;
  @JsonKey(name: 'zone_index')
  final int zoneIndex;
  @JsonKey(name: 'range_min')
  final num? rangeMin;
  @JsonKey(name: 'range_max')
  final num? rangeMax;
  @JsonKey(name: 'range_unit')
  final String rangeUnit;
  @JsonKey(name: 'duration_s')
  final num durationS;
  final num percent;

  Map<String, dynamic> toJson() => _$ZoneToJson(this);
}

@JsonSerializable()
class TimeseriesPoint {
  const TimeseriesPoint({
    this.timestamp,
    this.distance,
    this.heartRate,
    this.speed,
    this.adjustedPace,
    this.cadence,
    this.altitude,
    this.power,
  });

  factory TimeseriesPoint.fromJson(Map<String, dynamic> json) =>
      _$TimeseriesPointFromJson(json);

  final num? timestamp;
  final num? distance;
  @JsonKey(name: 'heart_rate')
  final num? heartRate;
  final num? speed;
  @JsonKey(name: 'adjusted_pace')
  final num? adjustedPace;
  final num? cadence;
  final num? altitude;
  final num? power;

  Map<String, dynamic> toJson() => _$TimeseriesPointToJson(this);
}

@JsonSerializable()
class Segment {
  const Segment({
    required this.segName,
    required this.lapIndex,
    required this.lapType,
    required this.distanceM,
    required this.distanceKm,
    required this.durationS,
    required this.durationFmt,
    required this.paceFmt,
    this.mode,
    this.avgPace,
    this.adjustedPace,
    this.avgHr,
    this.maxHr,
    this.avgCadence,
    this.avgPower,
    this.ascentM,
    this.descentM,
  });

  factory Segment.fromJson(Map<String, dynamic> json) =>
      _$SegmentFromJson(json);

  @JsonKey(name: 'seg_name')
  final String segName;
  final int? mode;
  @JsonKey(name: 'lap_index')
  final int lapIndex;
  @JsonKey(name: 'lap_type')
  final String lapType;
  @JsonKey(name: 'distance_m')
  final num distanceM;
  @JsonKey(name: 'distance_km')
  final num distanceKm;
  @JsonKey(name: 'duration_s')
  final num durationS;
  @JsonKey(name: 'duration_fmt')
  final String durationFmt;
  @JsonKey(name: 'avg_pace')
  final num? avgPace;
  @JsonKey(name: 'pace_fmt')
  final String paceFmt;
  @JsonKey(name: 'adjusted_pace')
  final num? adjustedPace;
  @JsonKey(name: 'avg_hr')
  final int? avgHr;
  @JsonKey(name: 'max_hr')
  final int? maxHr;
  @JsonKey(name: 'avg_cadence')
  final int? avgCadence;
  @JsonKey(name: 'avg_power')
  final num? avgPower;
  @JsonKey(name: 'ascent_m')
  final num? ascentM;
  @JsonKey(name: 'descent_m')
  final num? descentM;

  Map<String, dynamic> toJson() => _$SegmentToJson(this);
}

@JsonSerializable()
class ActivityDetailResponse {
  const ActivityDetailResponse({
    required this.activity,
    required this.laps,
    required this.segments,
    required this.zones,
    required this.timeseries,
  });

  factory ActivityDetailResponse.fromJson(Map<String, dynamic> json) =>
      _$ActivityDetailResponseFromJson(json);

  final Activity activity;
  final List<Lap> laps;
  final List<Segment> segments;
  final List<Zone> zones;
  final List<TimeseriesPoint> timeseries;

  Map<String, dynamic> toJson() => _$ActivityDetailResponseToJson(this);
}

File: src/wechat-miniprogram/types/activity.ts
// 活动 wire 类型 —— 与后端 /api/{user}/activities 响应及 Web 端 frontend/src/api.ts
// 的 Activity 契约保持一致（字段子集，今日页/活动页用到的部分）。

export interface Activity {
  label_id: string;
  name: string | null;
  sport_type: number;
  /** 可为 null：后端（Go/Python）对个别活动（如未标注运动类型的历史数据）返回 null */
  sport_name: string | null;
  /** 上海本地 YYYY-MM-DD（后端在序列化前已转上海时区） */
  date: string;
  distance_m: number;
  distance_km: number;
  duration_s: number;
  duration_fmt: string;
  avg_pace_s_km: number | null;
  pace_fmt: string;
  avg_hr: number | null;
  max_hr: number | null;
  avg_cadence: number | null;
  calories_kcal: number | null;
  training_load: number | null;
  vo2max: number | null;
  train_type: string | null;

  // —— 以下为详情页字段（list 端点可能不返回，故可选）——
  ascent_m?: number | null;
  aerobic_effect?: number | null;
  anaerobic_effect?: number | null;
  temperature?: number | null;
  humidity?: number | null;
  feels_like?: number | null;
  wind_speed?: number | null;
  feel_type?: number | null;
  sport_note?: string | null;
  pauses?: Pause[];
  commentary?: string | null;
  commentary_generated_by?: string | null;
  commentary_generated_at?: string | null;
}

/** 手表暂停区间。timestamp 为原始 COROS 厘秒；换算方式与 timeseries 一致。 */
export interface Pause {
  start_ts: number | null;
  end_ts: number | null;
  type: number | null;
}

/** 自动公里分段（lap_type 'autoKm'）或力量段（'type2'）。 */
export interface Lap {
  lap_index: number;
  lap_type: string;
  distance_m: number | null;
  distance_km: number | null;
  duration_s: number | null;
  duration_fmt: string;
  avg_pace: number | null;
  pace_fmt: string;
  adjusted_pace: number | null;
  avg_hr: number | null;
  max_hr: number | null;
  avg_cadence: number | null;
  avg_power: number | null;
  ascent_m: number | null;
  descent_m: number | null;
}

export interface Segment extends Lap {
  seg_name: string;
  mode: number | null;
}

/** 手表上报区间（Go）或校准区间（Python），percent 为该区间时长占比。 */
export interface Zone {
  zone_type: string;
  zone_index: number;
  range_min: number | null;
  range_max: number | null;
  range_unit: string | null;
  duration_s: number | null;
  percent: number | null;
}

export interface TimeseriesPoint {
  timestamp: number | null;
  distance: number | null;
  heart_rate: number | null;
  speed: number | null;
  adjusted_pace: number | null;
  cadence: number | null;
  altitude: number | null;
  power: number | null;
  gps_lat: number | null;
  gps_lon: number | null;
}

export interface LinkedScheduledWorkout {
  id: number;
  abandoned_by_promote_at: string | null;
}

export interface ActivityStrideTrainingLoad {
  label_id: string;
  activity_date: string;
  sport: string | null;
  session_class: string | null;
  algorithm_version: number;
  calibration_id: number | null;
  cardio_load_raw: number | null;
  cardio_tss: number | null;
  external_tss: number | null;
  high_intensity_tss: number | null;
  mechanical_load: number | null;
  subjective_internal_load: number | null;
  training_dose: number | null;
  training_dose_source: string | null;
  cardio_coverage: number;
  external_coverage: number;
  high_intensity_coverage: number;
  coverage_status: string;
  load_confidence: string | null;
  excluded_from_pmc: boolean;
  reasons: string[];
}

export interface ActivityDetailResponse {
  activity: Activity;
  stride_training_load?: ActivityStrideTrainingLoad | null;
  laps: Lap[];
  segments: Segment[];
  zones: Zone[];
  /** 仅在请求 include=timeseries 时返回；否则缺失/空数组。 */
  timeseries?: TimeseriesPoint[];
  linked_scheduled_workout?: LinkedScheduledWorkout | null;
}

export interface ActivityMonthlySummary {
  activity_count: number;
  total_run_km: number;
  run_duration_s: number;
  duration_s: number;
}

export interface ActivitiesListResponse {
  total: number;
  offset: number;
  limit: number;
  activities: Activity[];
  /** 本页可见各上海月份的聚合（键为 YYYY-MM），可选——旧后端可能不返回 */
  monthly_summaries?: Record<string, ActivityMonthlySummary>;
}
